SCClassLibrary/External/Ableton/LinkClock.sc

```diff
@@ -0,0 +1,77 @@
+// clock with Ableton Link synchronization
+
+LinkClock : TempoClock {
+
+	var <>tempoChanged;
```

Comment:
If I understand correctly we want an event scheduled at time `t0` to be executed at time `t0 - aheadOfTime`. We should have `thisThread.seconds == t0` at execution and then maybe another value `thisThread.realSeconds == t0 - aheadOfTime`.

Also `sendBundle` use `thisThread.seconds` to [compute the OSC time tag](https://github.com/supercollider/supercollider/blob/ce66ee973455719a32eb055e3c2286a660622dea/lang/LangPrimSource/OSCData.cpp#L451), so there are two solutions if we want the sound to be really effective at `t0`:

* Make `sendBundle` use `thisThread.realSeconds` instead, and the usage would be as you described.
* Let `sendBundle` use `thisThread.seconds`, and you should have
```supercollider
s.latency = 0
t.aheadOfTime = 0.2// whatever pre-computing time is needed to prevent late messages
```

I would personally opt for the second option since `latency` should really mean "how late do we want to hear the sound".

Reply:
> We can't set the server latency to zero, because other threads with other clocks may also schedule bundles, and they will be late.

I don't understand how it is an issue. We wan't LinkClock events to hit the speaker right on time so I wouldn't make a lot of sense to have other thread  producing sound on the same server but with a non-zero latency ? Plus we can always set the latency thread-wise with the `\latency` entry in Pbind.

> Let sendBundle use thisThread.seconds as usual, but when synchronized with the other link clocks, this value corresponds to a time t + aheadOfTime

This is also what I had in mind, and scheduling would happen like this:

1. an event is scheduled at time `t0`, with a LinkClock setting its internal clock to the other clocks time plus `aheadOfTime`
1. the event is executed at time `t0 - aheadOfTime` for the other clocks, so at time `t0` for LinkClock and thus `thisThread.seconds == t0`.
1. a bundle is created with a latency `latency`. Internally `sendBundle` compute the OSC time tag with the value `thisThread.seconds + latency == t0 + latency` and send it at `t0 - aheadOfTime`.
1. the audio hits the speaker precisely at  `t0 + latency`

This way `aheadOfTime` means "how far ahead do we *compute* events, and `latency` means "how far *behind* do we hear events", and they do not depend on each other (meaning changing `aheadOfTime` won't change the effective audio latency).

Reply:
>I don't understand how it is an issue. We wan't LinkClock events to hit the speaker right on time so I wouldn't make a lot of sense to have other thread producing sound on the same server but with a non-zero latency ? 

It is not always the case that all sounds are to be synced to a common clock. This is an aesthetic decision which we have to leave open to the user.

Also you have to remember that many bits of code can be combined that may not even know about clock sync.

>Plus we can always set the latency thread-wise with the \latency entry in Pbind.

Clocks are not only used in event streams, but in arbitrary schedulers. It would be a major design change to move the latency variable away from the server into all these processes. Also I don't think it makes too much sense to do that, because after all it is the server who is late (it could be the clocks' business, but also that is debatable).

>This way `aheadOfTime` means "how far ahead do we *compute* events, and `latency` means "how far *behind* do we hear events", and they do not depend on each other (meaning changing `aheadOfTime` won't change the effective audio latency).

I am not sure I understand the implementation you suggest, but yes, this is what I meant.

Reply:
thanks! why is the `IDEF` necessary? https://github.com/smiarx/supercollider/commit/7357216c4074640c86e23784f0b1ef0f022938ff#diff-862389468167bb46f32e36801276b4cbR804

Reply:
Not necessary, it's just to avoid unnecessary calculations when Link is deactivated at compile time.

Reply:
ok, makes sense.

Reply:
Yes with the example in the helpfile and
```
s.latency = 0;
l.aheadOfTime = 0.2;
```
I get an average latency of <1ms between the two programs.